designs/0100-reduce_sum.md

```diff
@@ -0,0 +1,214 @@
+- Feature Name: Reduce Sum Parallelization
+- Start Date: 2020-03-10
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+The reduce_sum function is an attempt to add a parallelization utility that is much easier to use than map-rect without requiring a substantial changes to the autodiff stack.
+
+# Motivation
+[motivation]: #motivation
+
+The overall goal of reduce_sum is to make it easier for users to parallelize their models by streamlining how arguments are handled, hiding work scheduling, and making it more difficult to program something that will accidentally be inefficient.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+reduce_sum is effectively a parallel-for combined with a sum.
+
+For a set of input arguments, ```args0, args1, args2, ...``` and a scalar function ```f```, reduce_sum computes the scalar:
+
+```f(args0) + f(args1) + f(args2) + ...```
+
+In terms of probabilistic models, this is useful when there are N terms in a likelihood that combine multiplicatively. In this case, computing the log likelihood means computing the sum of the log of each of these terms. If a function ```f``` can be written that computes an individual log likelihood, then reduce_sum can be used to compute them all in parallel and accumulate the results.
+
+The actual meaning of N is flexible here. In a regression model, N might be the N different rows of the input dataframe, each which can be evaluated separately. In a hierarchical model, N might correspond to different groups -- perhaps there is some complex shared calculation for each group but these can all be done in parallel.
+
+reduce_sum is not useful, however, when there are dependencies between these likelihood terms. This can happen, for instance, if there were N terms in a Gaussian process likelihood. reduce_sum will not be useful for accelerating this.
+
+The signature we settled on for reduce_sum (which balances the different tradeoffs described in the Summary) is:
+
+```
+real reduce_sum(F func, T[] sliced_arg, int grainsize, T1 arg1, T2 arg2, ...)
+```
+
+where the function ```func``` has a signature like:
+
+```
+real func(int start, int end, T[] subset_sliced_arg, T1 arg1, T2 arg2, ...)
+```
+
+The array ```sliced_arg``` in the reduce_sum call represents the list of calculations to perform and sum.
```

Comment:
I don't understand what you mean by _list of calculations to perform and sum_.  in the example below, the 2nd argument to `reduce_sum` is variable `Y`

Reply:
> is this the vector which reduce_sum will slice up instead of making the user do the work ahead of time?

Yes, but again I think I wrote this a little weirdly here so I'll try to redo with a rewrite.